source/opt/type_manager.cpp

```diff
@@ -940,7 +968,8 @@ void TypeManager::AttachDecoration(const Instruction& inst, Type* type) {
   if (!IsAnnotationInst(opcode)) return;
 
   switch (opcode) {
-    case spv::Op::OpDecorate: {
+    case spv::Op::OpDecorate:
+    case spv::Op::OpDecorateId: {
```

Comment:
Some explanation is need for why this will work. The decorations in the type were written assuming that the extra operands are literals, and not ids. By doing this you are losing this distinction because the opcode is not part of the "data".

Some comment might be needed to explained how people looking at the decorations in the type can know how to interpret the operands. Can this be done based solely on the decoration?

Reply:
I believe there is a theoretical bug. However, I could be wrong if there is a detail in the spec I don't know about.

The decoration is stored as a vector of integers. This is currently interpreted as the first element is the decoration id followed by a series of immediate values. This this change that is no longer always true. It could be a decoration id followed by a series of ids.  How can people reading the decorations associated with the type know the difference?

A potential problem. Suppose there are two types that are the exact same except that one has a decoration represented using OpDecorate and another with OpDecorateId:

```
OpDecorate %T1 SomeDecoration 100
OpDecorateId %T2 SomeDecoration %100
```

When hashing `%T1` and `%T2`, their decorations will have the exact same integers: `{SomeDecoration, 100}`. They will then hash to the same value, and be considered the same type, which they are not.

I could see there being a convention, which I do not see in the spec, that a decoration is either always used with `OpDecorate` or `OpDecorateId`, so this will not happen. If that is the case, we will need a authoritative list of which decorations use one opcode and which one use another. That way code that accesses `Type::decorations()` will know how to interpret them. Or we could come up with another way of disinguishing them by adding the opcode to the list or adding a separate list like we do for member variables. I'm fine with either solution?

Reply:
I don't think any decoration can be used with both IDs and literals. Usually we have a separate enum for those cases (e.g. LocalSize and LocalSizeID).

Reply:
Is there an easy way in Spirv-tools to be able to tell the difference?

Reply:
If we are expecting people to look at the decoration to know if it came from opdecorateid or not, then a comment should be added to the code.

Reply:
I do attempt to validate this in the validator. That requires keeping it up-to-date, but hopefully that would happen.

Otherwise you'd have to parse the grammar and check the operands which is awkward.

Reply:
@s-perron it's not clear to me what commentary you're looking for here. If this was the validator I agree it would need something explaining "validated elsewhere", but for the optimiser we should be assuming valid code, and it's plain at least to me that decorating with either instruction would achieve the same goal.

The decoration itself can be separately decoded later in isolation of the instruction it was provided with, because the decoration decides what its operands mean, not the instruction. You never need to know which instruction it came with, you can just assume was specified with the right one.

Are you looking for a comment outlining something like that?

Reply:
We should have comments added here to describe the encoding of the decoration in the vector:

https://github.com/KhronosGroup/SPIRV-Tools/blob/35e5f1160ecd2b963682e14e869001a60160a0c4/source/opt/types.h#L222-L225

That should be good enough. I can't find any uses that need to be updated.

Reply:
Add something like

> The parameters to the decoration could be either literals or ids depending on the decoration value.